app/src/sharedTest/java/org/oppia/android/app/splash/SplashActivityTest.kt

```diff
@@ -1049,6 +1058,72 @@ class SplashActivityTest {
     }
   }
 
+  @Test
+  fun testSplashActivity_initialOpen_onboardingV2Enabled_routesToOnboardingActivity() {
+    initializeTestApplication(onboardingV2Enabled = true)
+
+    launchSplashActivityPartially {
+      intended(hasComponent(OnboardingActivity::class.java.name))
+    }
+  }
+
+  @Test
+  fun testSplashActivity_onboardingV2Enabled_profilePartiallyOnboarded_routesToIntroActivity() {
+    initializeTestApplication(onboardingV2Enabled = true)
+    profileTestHelper.addOnlyAdminProfileWithoutPin()
+    val profileId = ProfileId.newBuilder().setInternalId(0).build()
+    profileTestHelper.updateProfileType(profileId, ProfileType.SOLE_LEARNER)
+    profileTestHelper.markProfileOnboardingStarted(profileId)
+    val params = IntroActivityParams.newBuilder()
+      .setProfileNickname("Admin")
+      .build()
+
+    launchSplashActivityPartially {
+      intended(hasComponent(IntroActivity::class.java.name))
+      intended(hasProtoExtra(IntroActivity.PARAMS_KEY, params))
+      intended(hasProtoExtra(PROFILE_ID_INTENT_DECORATOR, profileId))
+    }
+  }
+
+  @Test
+  @RunOn(TestPlatform.ESPRESSO)
```

Comment:
Why can't this run on Robolectric?

Reply:
There seems to be a synchronizing issue with this test. I have updated the test code itself:
```kotlin
 @Test
  @RunOn(TestPlatform.ESPRESSO)
  fun testSplashActivity_onboardingV2Enabled_onboardedSoleLearnerProfile_routesToHomeActivity() {
    simulateAppAlreadyOnboarded()
    initializeTestApplication(onboardingV2Enabled = true)
    profileTestHelper.addOnlyAdminProfileWithoutPin()
    testCoroutineDispatchers.runCurrent()

    val profileId = ProfileId.newBuilder().setInternalId(0).build()
   monitorFactory.ensureDataProviderExecutes(
      profileTestHelper.updateProfileType(profileId, ProfileType.SOLE_LEARNER)
    )
    monitorFactory.ensureDataProviderExecutes(
      profileTestHelper.markProfileOnboardingStarted(profileId)
    )
    monitorFactory.ensureDataProviderExecutes(
      profileTestHelper.markProfileOnboardingEnded(profileId)
    )
    testCoroutineDispatchers.advanceUntilIdle()

    launchSplashActivityPartially{
      intended(hasComponent(HomeActivity::class.java.name))
    }
  }
```

`profileTestHelper.markProfileOnboardingEnded(profileId)` never executes, causing the test to run indefinitely due to never entering an idle state. 
On Espresso, this passes but ocassionally fails with:

```
Wanted but not invoked:
observer.onChanged(<Capturing argument>);
-> at org.oppia.android.testing.data.DataProviderTestMonitor.ensureNextResultIsPresent(DataProviderTestMonitor.kt:117)
Actually, there were zero interactions with this mock.

at org.oppia.android.testing.data.DataProviderTestMonitor.ensureNextResultIsPresent(DataProviderTestMonitor.kt:117)
at org.oppia.android.testing.data.DataProviderTestMonitor.waitForNextResult(DataProviderTestMonitor.kt:59)
```

For any random DataProvider in the test.

Reply:
This turned out to be a dataprovider update loop, and I have fixed it.

Reply:
I am actually very interested in what the specific problems were, and if this is something we can try to document on the data providers wiki page. Let's discuss further.